Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp

```diff
@@ -376,45 +393,81 @@ void JitArm64::boolX(UGeckoInstruction inst)
         if (inst.Rc)
           ComputeRC0(gpr.GetImm(a));
       }
-      else if (final_not || complement_b)
+      else if (is_ones)
       {
-        gpr.BindToRegister(a, a == j);
-        MVN(gpr.R(a), gpr.R(j));
+        if (final_not || complement_b)
+        {
+          gpr.BindToRegister(a, a == j);
+          MVN(gpr.R(a), gpr.R(j));
+        }
+        else if (a != j)
+        {
+          gpr.BindToRegister(a, false);
+          MOV(gpr.R(a), gpr.R(j));
+        }
         if (inst.Rc)
           ComputeRC0(gpr.R(a));
       }
       else
       {
-        if (a != j)
+        if (!complement_b)
         {
-          gpr.BindToRegister(a, false);
-          MOV(gpr.R(a), gpr.R(j));
+          gpr.BindToRegister(a, a == j);
+          AND(gpr.R(a), gpr.R(j), log_imm);
+          if (final_not)
+            MVN(gpr.R(a), gpr.R(a));
+        }
+        else
+        {
+          // No shorter instruction sequence is possible. Just materialize the
+          // immediate in a register as usual, so subsequent uses can leech off
+          // of it.
+          gpr.BindToRegister(a, (a == i) || (a == j));
+          BIC(gpr.R(a), gpr.R(i), gpr.R(j));
         }
         if (inst.Rc)
           ComputeRC0(gpr.R(a));
       }
     }
     else if (is_or)
     {
-      if (!is_zero)
+      if (is_ones)
       {
         gpr.SetImmediate(a, final_not ? 0 : 0xFFFFFFFF);
         if (inst.Rc)
           ComputeRC0(gpr.GetImm(a));
       }
-      else if (final_not || complement_b)
+      else if (is_zero)
       {
-        gpr.BindToRegister(a, a == j);
-        MVN(gpr.R(a), gpr.R(j));
+        if (final_not || complement_b)
+        {
+          gpr.BindToRegister(a, a == j);
+          MVN(gpr.R(a), gpr.R(j));
```

Comment:
If this is a bug, it must be a pre-existing bug, but... This shouldn't run for `complement_b`, should it? If we have `is_zero && complement_b`, that means `s` is an immediate zero, so the result should be zero regardless of what `b` is. (If the immediate zero had been in `b`, code earlier would have set `complement_b` to false.)

Reply:
We're doing `is_or`. There's three possibilities.

1. `final_not` with immediate in s: `~(0 | b) == ~b == ~j`
2. `final_not` with immediate in b: `~(s | 0) == ~s == ~j`
3. `complement_b` (immediate can only be in s): `0 | ~b == ~b == ~j`

In all cases, it works out to `~j`. This seems correct to me, but perhaps I'm missing something.

Reply:
Oh, right, it's OR and not AND. Yeah, ignore me then.